### PR TITLE
[ENT-587] More gracefully handle errors from hitting the Ecommerce API (and others)

### PR DESCRIPTION
**Pull request by UmanShahzad**
**Description:** This PR moves error handling logic for our external API clients into the clients themselves, rather than having try/except blocks all over the codebase.

**JIRA:** [ENT-587](https://openedx.atlassian.net/browse/ENT-587)

**Merge deadline:** Before the end of the week.

**Testing instructions:**

1. Since this is just a refactor and doesn't introduce any new functionality (but does add test coverage that was lost due to moving logic around), you can either rely on the tests or just give the course enrollment landing page a spin; see if the API calls are working properly and that in exceptional cases like missing certain data, we get the same behavior as last time.

**Merge checklist:**

- [ ] Check that the versions of the requirements in the `platform-master.in` file match edx-platform.
- [ ] New requirements are in the right place (`base.in` if only used in enterprise; in the correct `platform-****.in` files if they're hosted in edx-platform)
- [ ] Regenerate requirements with `make upgrade && make requirements` (and make sure to fix any errors).
  **DO NOT** just add dependencies to `requirements/*.txt` files.
- [ ] All reviewers approved
- [ ] CI build is green
- [ ] Version bumped
- [ ] Changelog record added
- [ ] Documentation updated (not only docstrings)
- [ ] Commits are (reasonably) squashed
- [ ] Translations are updated
- [ ] PR author is listed in AUTHORS

**Post merge:**
- [ ] Create a tag
- [ ] Check new version is pushed to PyPi after tag-triggered build is finished.
- [ ] Delete working branch (if not needed anymore)
- [ ] edx-platform PR (be sure to include edx-platform requirements upgrades that were present in this PR)

**Author concerns:** I do wonder if the defaults I return in some cases are appropriate. So please do a double-check there.

### CODE REVIEW

**Comment by edx-webhook** [comment]
Thanks for the pull request, @UmanShahzad! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

<a   href="https://openedx-webhooks.herokuapp.com/github/process_pr?number=182&repo=edx%2Fedx-enterprise"   rel="nofollow" >Create an OSPR issue for this pull request</a>.

**Comment by UmanShahzad** [comment]
I basically decided to cover all of our external API calls in `enterprise.views.py` to put enough work into this to match the estimated story points... :P

File: enterprise/api_client/discovery.py [hunk 1]
```diff
@@ -377,13 +383,16 @@ def _load_data(self, resource, default=DEFAULT_VALUE_SAFEGUARD, **kwargs):
...
-        return result or default_val
+        try:
+            return get_edx_api_data(
+                api_config=CatalogIntegration.current(),
+                resource=resource,
+                api=self.client,
+                **kwargs
+            ) or default_val
+        except HttpClientError as exc:
+            LOGGER.error('Failed to load data from resource %s with kwargs %s due to: %s', resource, kwargs, str(exc))
```
Comment (douglashall) [inline enterprise/api_client/discovery.py]: Can we change these to `LOGGER.exception` calls instead so we get the full stack trace in the logs?

Comment (UmanShahzad) [reply enterprise/api_client/discovery.py]: Yes, sure!

File: enterprise/api_client/discovery.py [hunk 2]
```diff
@@ -377,13 +383,16 @@ def _load_data(self, resource, default=DEFAULT_VALUE_SAFEGUARD, **kwargs):
...
-        )
-        return result or default_val
+        try:
+            return get_edx_api_data(
+                api_config=CatalogIntegration.current(),
+                resource=resource,
+                api=self.client,
+                **kwargs
+            ) or default_val
+        except HttpClientError as exc:
```
Comment (douglashall) [inline enterprise/api_client/discovery.py]: Does this need to be the `(ConnectionError, SlumberBaseException, Timeout)` trifecta like you have in the [ecommerce](https://github.com/edx/ecommerce/pull/1468/files#diff-0aa9e695700bf5edb4df6ef6ac088d71R87) PR?

Comment (UmanShahzad) [reply enterprise/api_client/discovery.py]: I'm honestly not sure about it. @haikuginger has more knowledge about Slumber -- he would know.

File: enterprise/api_client/discovery.py [hunk 3]
```diff
@@ -411,3 +420,13 @@ def __init__(self):
                 raise ImproperlyConfigured(_("The configured CatalogIntegration service user does not exist."))
         else:
             raise ImproperlyConfigured(_("There is no active CatalogIntegration."))
+
+    @classmethod
```
Comment (douglashall) [inline enterprise/api_client/discovery.py]: I don't think you want the `@classmethod` decorator here.

Comment (UmanShahzad) [reply enterprise/api_client/discovery.py]: I wanted to encapsulate handling the `ImproperlyConfigured` error into this function if possible to make the code in our views easy to read. For that, I needed to instantiate it here. Or do you mean make it a static method?

Comment (douglashall) [reply enterprise/api_client/discovery.py]: Ah, I see now. Makes sense. Thanks for the clarification.

File: enterprise/api_client/discovery.py (hunk 2 above)
Comment (douglashall) [reply enterprise/api_client/discovery.py]: @haikuginger What do you think? Seems like we should at least change this to `SlumberBaseException`.

Comment (haikuginger) [reply enterprise/api_client/discovery.py]: @douglashall, as far as I can tell, Slumber doesn't catch those other two, so let's include them.

Comment (UmanShahzad) [reply enterprise/api_client/discovery.py]: :+1:

**Comment by UmanShahzad** [comment]
@haikuginger @douglashall Added more exceptions for catching. Anything in particular that's left here?

**Comment by UmanShahzad** [comment]
@douglashall @brittneyexline Look good to you guys? I'd like to rebase the programs landing page on this to benefit from it if possible before we merge that.

**Pull request merged**